intra/protect/sockopt.go

```diff
@@ -24,28 +24,38 @@
 package protect
 
 import (
+	"net"
 	"syscall"
 
 	"github.com/celzero/firestack/intra/log"
 )
 
 const (
-	defaultKeepIdle = 120
-	defaultKeepCnt = 5
+	defaultKeepIdle  = 120
+	defaultKeepCnt   = 5
 	defaultKeepIntvl = 1
 )
 
-func SetKeepAliveConfig(network, addr string, c syscall.RawConn) error {
-	return c.Control(func(fd uintptr) {
-		sock := int(fd)
-		if err := syscall.SetsockoptInt(sock, syscall.IPPROTO_TCP, syscall.TCP_KEEPIDLE, defaultKeepIdle); err!=nil {
-			log.E("set TCP_KEEPIDLE failed: %v", err)
+func TrySetKeepAliveConfig(c Conn) {
+	if tcpConn, ok := c.(*net.TCPConn); ok {
+		rawConn, err := tcpConn.SyscallConn()
```

Comment:
I'm confused. Why does it not panic when c==nil ? @ignoramous

Reply:
A *safe* typecast where `ok` is set to `false` when the typecast fails, instead of the runtime panicking. https://archive.fo/pqM3K / https://medium.com/gopher-time/safe-casting-in-go-the-power-of-the-ok-idiom-fce5fc117040

Another common idiom to safe typecast in golang are the `switch x := y.(type)` statements.